base/shared/src/main/scala/scalaz/data/lazy.scala

```diff
@@ -0,0 +1,110 @@
+package scalaz
+package data
+
+import Predef._
+import scalaz.prop.IsCovariant
+
+import scala.AnyVal
+import scalaz.tc.{ instanceOf, Comonad, ComonadClass, Monad, MonadClass, Traversable, TraversableClass }
+
+sealed abstract class LazyModule {
+  type Lazy[A]
+
+  def now[A](a: A): Lazy[A]
+
+  def delay[A](run: () => Lazy[A]): Lazy[A]
+
+  def eval[A](thunk: Lazy[A]): A
+
+  final object Eval {
+    def unapply[A](thunk: Lazy[A]): Evaluated[A] = new Evaluated[A](eval(thunk))
+  }
+}
+
+final class Evaluated[A] private[data] (private val value: A) extends AnyVal {
+  def isEmpty: Boolean = false
+  def get: A           = value
+}
+
+object LazyModule {
+
+  final implicit def isCovariant[A]: IsCovariant[Lazy[?]] = IsCovariant.unsafeForce // TODO: do better
+
+  implicit val monad: Monad[Lazy] = instanceOf(
+    new MonadClass[Lazy] {
+      override def pure[A](a: A): Lazy[A] = Lazy.now(a)
+      override def flatMap[A, B](ma: Lazy[A])(f: A => Lazy[B]): Lazy[B] =
+        Lazy.delay(() => f(Lazy.eval(ma)))
+    }
+  )
+
+  implicit val traversable: Traversable[Lazy] = instanceOf(
+    new TraversableClass[Lazy] {
+      override def map[A, B](ma: Lazy[A])(f: A => B): Lazy[B] =
+        Lazy.delay(() => Lazy.now(f(Lazy.eval(ma))))
+      override def foldLeft[A, B](fa: Lazy[A], z: B)(f: (B, A) => B): B        = f(z, Lazy.eval(fa))
+      override def foldRight[A, B](fa: Lazy[A], z: => B)(f: (A, => B) => B): B = f(Lazy.eval(fa), z)
+    }
+  )
+
+  implicit val comonad: Comonad[Lazy] = instanceOf(
+    new ComonadClass[Lazy] {
+      override def copoint[A](fa: Lazy[A]): A = Lazy.eval(fa)
+      override def map[A, B](ma: Lazy[A])(f: A => B): Lazy[B] =
+        Lazy.delay(() => Lazy.now(f(Lazy.eval(ma))))
+      override def cobind[A, B](fa: Lazy[A])(f: Lazy[A] => B): Lazy[B] =
+        Lazy.delay(() => Lazy.now(f(fa)))
+    }
+  )
+}
+
+private[data] object LazyModuleImpl extends LazyModule {
+
+  override type Lazy[A] = scala.Any
+
+  override def now[A](a: A): Lazy[A] = a
+
+  override def delay[A](a: () => Lazy[A]): Lazy[A] = new Delay[A](a)
+
+  override def eval[A](thunk: Lazy[A]): A = thunk match {
+    case d: Delay[A] @scala.unchecked => d.eval
+    case a: A @scala.unchecked        => a
+  }
+
+  private final class Delay[A](
+    @scala.volatile
+    private var run: () => Lazy[A]
+  ) {
+
+    private var value: A = _
+
+    def eval: A =
+      if (run eq null)
+        value
+      else runEval
+
+    private def runEval: A = this.synchronized {
```

Comment:
I believe that this cannot be observed through the LazyModule API, so parametricity should be preserved. The semantic of Lazy.eval should be the same as pattern matching in Haskell: it evaluate a thunk to weak-head normal form (no matter the number of nested lazy layers).

Reply:
It is but limited to the Haskell thunk semantic. Also thread-safe, and, for the most outer layer, it guarantee at-most-once evaluation.

Reply:
That doesn't appear to be true; if I call `.value` on a  `Lazy[Lazy[A]]`, it will force *both* layers at once, no?

Reply:
@edmundnoble yes, it will force both layer, but there is no way to observe that, because `A` is valid runtime representation for `Lazy[A]` (now `Eval[A]`).

Reply:
@jbgi well except with bottom, is this case you can observe, right. I can leave with that, as there no use cases for pealing off only one layer of `Lazy[Lazy[A]]' since the only thing you do with the result is to peal off more layers.